### DISCUSSION TITLE
Control order of the annotations to prevent

### DISCUSSION THREAD

**Discussion by llllkatjallll** [google/model-viewer#3029] [Q&A]
Hi!
I was wondering if it is possible to change the order in which the annotations are being shown on my html canvas?
If I understand it right, currently the annotation with a smaller distance to  the camera is always placed on top. 
In some situations it makes sense, but sometimes this leads to things like this:
<img width="951" alt="chrome_vvWNNx2uVL" src="https://user-images.githubusercontent.com/38395689/144892441-022eb5ec-5bda-4c0c-83bf-c6eb1470e796.png">
You can see this example here: https://prefrontalcortex.de/labs/model-viewer/hotspot-order/

I thought that by re-ordering the annotations in html or by changing the z-index on click/hover with JS would overwrite previous order but none of this has worked.

Are there any other possibilities to control the annotation order?

**Comment by elalish** [comment]
Huh, yeah that's not  great. we don't really control ordering internally since we have no idea what content you might have  put in. I'm surprised that the z-index didn't fix  it; maybe just set the z-index of the annotations to be in front of the buttons?

**Reply by llllkatjallll** [reply]
It didn't work eather.
I did more tests and it seems like hotspots that are attached to model-viewer model are ignoring the z-index of another attached hotsopots (but they respect z-index of elements that are not attached to the model.

Here is an example:
Yellow and blue hotspots are not attached and have z-index 1(blue) and 4(yellow). Attached hotspots have z-index 2(red) and 3(green). (All hotspots are children of the model-viewer element)
<img width="300" alt="chrome_ixGT3nwOL0" src="https://user-images.githubusercontent.com/38395689/145040591-d44c3ab4-ccb5-4d64-8175-c72f40c61ac0.png">
<img width="300" alt="chrome_7kfGALgICW" src="https://user-images.githubusercontent.com/38395689/145040612-8b3b57c1-d284-44b4-96e0-97d4ba799e7e.png">

Red and green hotspots are placed over the blue hotspot and under the yellow one, what is right, but green one should be over the red one, what is wrong :(

You can try it here: https://prefrontalcortex.de/labs/model-viewer/hotspot-order/
There are some other tests and variations too.
<img width="647" alt="chrome_wwv2ZwJ1Pc" src="https://user-images.githubusercontent.com/38395689/145019772-a3fae0fc-93bb-4e45-acf9-31c8b9ea3282.png">

**Reply by llllkatjallll** [reply]
Edit: it does not always respect z-index of hotspots that are not attached do the model: 
Here the green hotspot should be always on top because of z-index 4. But if I rotate the model, it is sometimes under the blue hotspot(z-index 2) and sometimes over it.
![ezgif-2-760727989677](https://user-images.githubusercontent.com/38395689/145062672-cfc60406-0990-4834-bf2e-c2dcff6a1df3.gif)

You can reproduce it if you select the last button in the example I posted in my last comment and rotate the model: 
<img width="645" alt="chrome_N0SCMJf22V" src="https://user-images.githubusercontent.com/38395689/145063635-4eafed5f-5c34-444d-a32d-962cdf18bba2.png">

**Reply by elalish** [reply]
Thanks for the detail; I must admit I do not understand why it's behaving the way it is, but I'm also a little shaky on how z-index is supposed to work anyway. Would you be interested in testing some of the internals here? https://github.com/google/model-viewer/blob/master/packages/model-viewer/src/three-components/Hotspot.ts

We wrap each hotspot in a couple of `<div>`s, so maybe that's  somehow  confusing things. PRs certainly welcome.
